### DISCUSSION TITLE
Grafana isn't loading it's config

### DISCUSSION THREAD

**Discussion by j-mie** [prometheus-operator/kube-prometheus#806] [Q&A] [answered]
Our config looks a little like:

```jsonnet
local kp =
  (import 'kube-prometheus/kube-prometheus.libsonnet') +
  (import 'kube-prometheus/kube-prometheus-managed-cluster.libsonnet') +
  {
    grafana+:: {
      config: {
        sections: {
          server: {
            root_url: 'https://grafana',
            enable_gzip: true,
          }
        }
      }
    }
  }
```

When we build this it modifies the `grafana-config.yaml` file. Changing the config to the example config in the README:
```jsonnet
      grafana+:: {
        config: { // http://docs.grafana.org/installation/configuration/
          sections: {
            "auth.anonymous": {enabled: true},
          },
        },
      },
```

Deletes our custom settings but leaves the file in the same format:

![image](https://user-images.githubusercontent.com/1857320/100098602-2fdcd980-2e56-11eb-8d8d-b597c8c274df.png)

But Grafana does not seem to load this custom config.

Looking in the grafana library I noticed:
```jsonnet
  grafana+: {
    [if std.length($._config.grafana.config) > 0 then 'config']:
      {
        apiVersion: 'v1',
        kind: 'Secret',
        metadata: {
          name: 'grafana-config',
          namespace: $._config.namespace,
        },
        type: 'Opaque',
        data: {
                'grafana.ini': std.base64(std.encodeUTF8(std.manifestIni($._config.grafana.config))),
              } +
              if $._config.grafana.ldap != null then { 'ldap.toml': std.base64(std.encodeUTF8($._config.grafana.ldap)) } else {},
      },
```

So is the README in this repo out of date? Or am I missing something on our config that's causing the file not to get loaded?

**Comment by brancz** [comment]
What is it that you are trying to do exactly, as in what was the expected result that didn't happen?

Off the top of my head, I'm guess you probably want to merge configuration instead of overriding, which you can do by using `+:` in jsonnet. So that would be something along the lines of:

```jsonnet
      grafana+:: {
        config+: {
          sections+: {
            "auth.anonymous": {enabled: true},
          },
        },
      },
```

**Reply by j-mie** [reply]
Sorry - I'll be more clear

```jsonnet
    grafana+:: {
      config: {
        sections: {
          server: {
            root_url: 'https://grafana',
            enable_gzip: true,
          },
          auth: {
            oauth_auto_login: true,
            disable_login_form: true,
          },
          'auth.anonymous': {
            enabled: false
          },
          'auth.basic': {
            enabled: false
          },
          'auth.github': {
              enabled: true,
              allow_sign_up: true,
              client_id: 'c',
              client_secret: 'c',
              scopes: "user:email,read:org",
              auth_url: 'g',
              token_url: 'g',
              api_url: 'g',
              allowed_organizations: 'o',
          },
          smtp: {
            enabled: true,
            host: 'h',
            skip_verify: true,
            from_address: "e",
            from_name: "n",
          },
        },
      },
    },
```

Our Grafana config looks like this, it's generating a file called `grafana-config.yml`, when we remove all of the config and regenerate it we see the yml file modified as expected - but regardless of what's in our config we do not see these settings loaded by the grafana server, that is the problem we're facing.

**Reply by brancz** [reply]
That's odd, can you share your grafana logs?

**Reply by j-mie** [reply]
The logs looked fine, I finally traced it down - our Grafana config was in the wrong scope!!!

**Comment by j-mie** [comment, marked as answer]
Grafana config wasn't scoped correctly, should have been in _config, troubleshooted by using [std.trace](https://jsonnet.org/ref/stdlib.html)
